### PR TITLE
React plot

### PR DESCRIPTION
**Pull request by cheekw**
- cleaned up matrix a bit
- added config typing
- react migration with functional bar, pie, histogram, scatter

### CODE REVIEW

File: src/components/matrix/grid-button.tsx [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+export type GridMoveButton = {
+  onClick: MouseEventHandler;
+  rowHeaderWidth: number;
+};
+
+/**
+ * SVG Icon for button from https://icons.getbootstrap.com/
+ */
+export const GridLeftButton = ({ onClick, rowHeaderWidth }: GridMoveButton): JSX.Element => {
```
Comment (cheekw) [inline src/components/matrix/grid-button.tsx:11]: @RFSH moved these buttons to a separate file to make the grid component cleaner

File: src/hooks/matrix.ts [hunk 2]
```diff
@@ -160,7 +162,7 @@ type MatrixData = {
   /**
    * matrix configurations
    */
-  config: any;
+  config: MatrixDefaultConfig | null;
```
Comment (cheekw) [inline src/hooks/matrix.ts:165]: added typing @RFSH

File: src/models/matrix-config.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
+export type MatrixDefaultConfig = {
```
Comment (cheekw) [inline src/models/matrix-config.ts:1]: Matrix config typing @RFSH

File: package.json [hunk 4]
```diff
@@ -26,6 +29,7 @@
   },
   "dependencies": {
     "@isrd-isi-edu/chaise": "0.0.8",
+    "plotly.js-cartesian-dist-min": "^2.18.2",
```
Comment (cheekw) [inline package.json:32]: Added cartesian to support plotly plots 

@jrchudy

File: src/utils/string.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+};
+
+/**
+ * Creates the link bemplateased on the given markdown pattern and templateParam
+ *
+ * @param markdownPattern pattern used to create link 
+ * @param templateParam params used to replace values in the pattern to create the link
+ * @returns
+ */
+export const createLink = (markdownPattern: string, templateParam?: any): string => {
```
Comment (cheekw) [inline src/utils/string.ts:112]: Created some helper functions that were often getting used in plot.app.js

@jrchudy

File: types/plotly.js-cartesian-dist-min.d.ts [hunk 6]
```diff
@@ -0,0 +1,4 @@
+declare module 'plotly.js-cartesian-dist-min' {
```
Comment (cheekw) [inline types/plotly.js-cartesian-dist-min.d.ts:1]: for typing

File: src/models/matrix-config.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  xURL: string;
+  yURL: string;
+  zURL: string;
+  xysURL: string;
+  xFacetColumn: string;
+  yFacetColumn: string;
+  zFacetColumn: string;
+  xSource: AxisSource;
```
Comment (RFSH) [inline src/models/matrix-config.ts:9]: Let's use `any` for the `xSource`, `ySource`, and `zSource`. Its value could be a string or array of elements. The typing should be something that we're eventually going to add to ermrestjs.

**Review by RFSH — APPROVED** [review]
The matrix changes look good to me.

**Review by jrchudy — APPROVED** [review]
This looks really good. I tested each example in the github issue I created and can see each working as I expect. The only inconsistency is the location of the legend by default for pie charts, but as we discussed on Wednesday that seems to be something in plotly that has changed.

Only question I had was regarding using the `plotly.js-cartesian-dist-min` library. Was this because not all types we want to support  were supported in the basic version of the library? I imagine violin might be one type that wasn't supported in the basic library.

**Pull request merged**